txpipe/extensions/twopoint_scia.py

```diff
@@ -0,0 +1,502 @@
+from ..base_stage import PipelineStage
+from ..twopoint import TXTwoPoint 
+from ..data_types import HDFFile, ShearCatalog, TomographyCatalog, RandomsCatalog, FiducialCosmology, SACCFile, PhotozPDFFile, PNGFile, TextFile
+from ..utils.calibration_tools import apply_metacal_response, apply_lensfit_calibration 
+from ..utils.calibration_tools import read_shear_catalog_type
+import numpy as np
+import random
+import collections
+import sys
+
+# This creates a little mini-type, like a struct,
+# for holding individual measurements
+Measurement = collections.namedtuple(
+    'Measurement',
+    ['corr_type', 'object', 'i', 'j'])
+
+SHEAR_SHEAR = 0
+SHEAR_POS = 1
+POS_POS = 2
+SHEAR_POS_SELECT = 3
+
+class TXSelfCalibrationIA(TXTwoPoint):
+    """
+    This is the subclass of the Twopoint class that will handle calculating the
+    correlations needed for doing the Self-calibration Intrinsic alignment
+    estimation. 
+
+    It requires estimating 3d two-point correlations. We calculate the 
+    galaxy - galaxy lensing auto-correlation in each source bin.
+    We do this twice, once we add a selection-function, such that we only selects 
+    the pairs where we have the shear object in front of the object used for density. 
+    Once without imposing this selection funciton. 
+    """
+    name = 'TXSCIA'
+    inputs = [
+        ('shear_catalog', ShearCatalog),
```

Comment:
I'm working on a new version of TXTwoPoint that might make a simpler basis for this, but for now that isn't important.

Reply:
Yeah, I will probably have to update the extension at that point!